### PR TITLE
Fixed #24887 -- Removed one-arg limit from models.aggregate

### PR DESCRIPTION
**Pull request by gchp**

### CODE REVIEW

**Comment by charettes** [comment]
Could you add a test for this? Also commit message should be in the past tense: `Removed one-arg ...`.

**Comment by coldmind** [comment]
@charettes https://code.djangoproject.com/ticket/24887#comment:5

**Comment by gchp** [comment]
@charettes I've updated this based on the comments on the ticket. I'm a little unsure of how to test this, though. How can I make `source_expressions` have more than just one element?

**Comment by jarshwah** [comment]
`GROUP_CONCAT` exists for sqlite and mysql (which takes multiple args), but doesn't for postgres or oracle. So you can invent a multiple-arg aggregate:

```python
class MyMax(Max):
    def as_sql(...):
        self.set_source_expressions(
            self.get_source_expressions[0:1]
        )
        return super(MyMax, self).as_sql(...)

Model.objects.annotate(MyMax('field', 'other_field')  # fail due to no alias
Model.objects.annotate(max_field=MyMax('field', 'other_field')  # pass
```

The above will let you pass two source expressions all the way through, but then remove the second just before constructing the sql output. I think that'll do the trick.

**Comment by gchp** [comment]
@jarshwah thanks! That did the trick. Just updated with a test.

**Comment by jarshwah** [comment]
Please also test that the lack of alias fails with a FieldError

**Comment by gchp** [comment]
@jarshwah the example you mentioned doesn't trigger the missing alias failure. Both source_expressions have the 'name' property

**Comment by gchp** [comment]
Actually, it seems like neither of the examples are getting past the 

`if source_expression.contains_aggregate and not summarize` line

File: django/db/models/aggregates.py [hunk 1]
```diff
@@ -15,13 +15,18 @@ class Aggregate(Func):
...
-            name = self.source_expressions[0].name
-            raise FieldError("Cannot compute %s('%s'): '%s' is an aggregate" % (
-                c.name, name, name))
+        for index, source_expression in enumerate(c.source_expressions):
+            if source_expression.contains_aggregate and not summarize:
+                expr = self.source_expressions[index]
+                if hasattr(expr, 'name'):
+                    name = expr.name
+                else:
+                    raise TypeError("Complex expressions require an alias")
```
Comment (jarshwah) [inline django/db/models/aggregates.py]: This should still be handled by `default_alias`. Something like:

```
    @property
    def default_alias(self):
        if len(self.source_expressions) == 1 and hasattr(self.source_expressions[0], 'name'):
            return '%s__%s' % (self.source_expressions[0].name, self.name.lower())
        raise TypeError("Complex expressions require an alias")
```

File: django/db/models/aggregates.py [hunk 2]
```diff
@@ -15,13 +15,18 @@ class Aggregate(Func):
...
+                name = self.source_expressions[index].name
+                raise FieldError("Cannot compute %s('%s'): '%s' is an aggregate" % (
```
Comment (jarshwah) [inline django/db/models/aggregates.py]: This error no longer makes sense with multiple-arg aggregates. You'll either need to join the output of all source expressions so the error produces:

> Cannot compute Sum(arg1, arg2): 'X' is an agggregate # (where X is either arg1 or arg2)

Or you'll need a different error message for the case of `len(args) > 1`.

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/4900.

**Pull request closed without merging**